README.md
# Azote

**Azote** is a GTK+ 3-based picture browser and a wallpaper setter, as the frontend to the [swaybg](https://github.com/swaywm/swaybg) 
(Sway/Wayland) and [feh](https://feh.finalrewind.org) (X windows) commands. During development I test it on sway, i3, 
Openbox, qtile and dwm, so I'm not sure if it works on others window managers.

![screenshot](http://nwg.pl/Lychee/uploads/big/2ef98871aea09679282675e942f153ed.png)

*Pictures above come from https://wallhaven.cc*

Also see [Azote in action on YouTube](https://youtu.be/Cjqr0LRL67I).

## Project assumptions

The most commonly used *desktop background browser and setter* is aimed at X windows, and does not work with 
[sway](https://swaywm.org). Since the `swaybg` command does everything we may need, it's enough to give it a GUI. 
In order not to limit the program usage to the single environment, Azote is also capable of using feh 
when running on i3, Openbox or other X11 window managers.

### Main features:

- works on Sway
- uses own, bigger thumbnails (240x135px)
- flips wallpapers horizontally
- splits wallpapers between 2 or more displays

## Usage

Select the folder your wallpapers are stored in. If it contains a lot of big pictures, it may take some time for
Azote to create thumbnails. It's being performed once per folder, unless you clear the `~/.azote/thumbnails` folder.

Most of the buttons seem to be self-explanatory, with a little help from their tooltip text. What may not be clear
at first is the `Apply selected picture to all screens` button. It applies unchanged
selected picture to all displays, regardless of whether they are currently connected/detected. It may be useful if you
often connect and disconnect displays. A shortcut to this feature is just to double click a thumbnail. It'll always 
use the 'fill' mode, however.

Azote, as well as feh, saves a batch file to your home directory. It needs to be executed in order to set the wallpaper 
on subsequent logins or reboot.

### sway

Edit your `~/.config/sway/config` file. Replace your current wallpaper settings, like:

```bash
output * bg /usr/share/backgrounds/sway/Sway_Wallpaper_Blue_1920x1080.png fill
```

with:

```bash
exec ~/.azotebg
```

### X window managers (i3, Openbox, dwm etc.)

You need to execute `~/.fehbg` from your window manager’s startup file.

**dwm note:**

If you start dwm from a script, it may look something like this:

```bash
# Statusbar loop
while true; do
   xsetroot -name "$( date +"%F %R" )"
   sleep 1m    # Update time every minute
done &

# Autostart section
~/.fehbg & 

exec dwm
```

## Versioning

Azote uses [Semantic Versioning 2.0.0](https://semver.org). 
For changes see [CHANGELOG](https://github.com/nwg-piotr/azote/blob/master/CHANGELOG).

## Installation

[![Packaging status](https://repology.org/badge/vertical-allrepos/azote.svg)](https://repology.org/project/azote/versions)

### Arch Linux

Install [azote](https://aur.archlinux.org/packages/azote) from AUR. 
For the development version install [azote-git](https://aur.archlinux.org/packages/azote-git).

### Debian & Ubuntu

Either download the .deb package or follow the instructions in [this link](https://software.opensuse.org//download.html?project=home%3AHead_on_a_Stick%3Aazote&package=azote) to add the repository and APT key.

If the repository & key are added then the package will be updated with the usual `apt update && apt upgrade` commands.

### openSUSE
azote is available in jubalhs home [repository](https://build.opensuse.org/package/show/home:jubalh/azote) on OBS.

```
zypper ar obs://home:jubalh
zypper ref
zypper in azote
```

### Void Linux

Binary package `azote` available in the Void repository. 

### Other Linux distributions:

Packagers wanted. Personally I only maintain Arch (AUR) and Void Linux packages.

**Dependencies:**

- python
- python-setuptools
- python-gobject
- python-pillow 
- gtk3
- feh 
- xorg-xrandr
- wmctrl

**Optional:** 

- python-send2trash: trash support

Please use assets from the [latest release](https://github.com/nwg-piotr/azote/releases/latest).

Seeing Arch [PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=azote) may be informative.

## Troubleshooting

### No pictures in thumbnails / display preview

As well thumbnails, as displays preview inherit from the Gtk.Button class. In case you don't see images inside them,
please make sure that button images are turned on in the `~/.config/gtk-3.0/settings.ini` file:

```bash
[Settings]
(...)
gtk-button-images=1
```

### Open selected picture button doesn't work

![Open image menu](http://nwg.pl/Lychee/uploads/big/156b73ac880bca02ed04bd1886404ba8.png)

**Azote v1.2.0 and below** - the button does not work at all.

**Azote v1.3.0 and above** - the button opens selected picture with feh, does not allow to choose a program.

The `/usr/share/applications/mimeinfo.cache` is probably missing from your system. Regenerate it:

```bash
$ sudo update-desktop-database
```

See https://specifications.freedesktop.org/desktop-entry-spec/0.9.5/ar01s07.html

### X11 / feh notice

The background color picker won't be available. You'll also be unable to select different modes 
*("scale", "max", "fill", "center", "tile")* for certain displays. The list of modes varies from what you see in Sway 
*("stretch", "fit", "fill", "center", "tile")*.


CHANGELOG
v1.5.0 (2019-10-12)
- Image menu: scale and crop the selected image to dimensions of a selected display

v1.4.0 (2019-09-15)
- move to trash button feature moved to the image menu, button removed, context (right click) menu added
- image button now opens the image menu
- settings button added: turns image button and thumbnail context menu on/off (as right-click menu is not 100% reliable
  on Sway). Default settings determined on 1st run, depending on the sway environment detection.
- menus' gravity adjusted
- way of checking if we're running sway altered to avoid crash on qtile

v1.3.1 (2019-08-21)
- double click on a thumbnail sets the wallpaper on all displays (mode = 'fill')
- main.py: some methods in classes replaced with static functions
- fix to the "Azote doesn't handle correctly file paths with spaces" issue #41

v1.3.0 (2019-07-29)
- fr_FR translation by HumanG33k added
- fixed "silent" crash when running Azote w/o send2trash module
- initial window size altered
- thumbnail creation progress bar added
- fallback for missing /usr/share/applications/mimeinfo.cache added

v1.2.0 (2019-07-07)
- issue #30 resolved: folder picker starts from the folder currently in use
- feature requested in issue #29 added: new button 'Apply to all displays'
- former 'open with feh' button now opens the menu to select from programs capable of opening the image

v1.1.2 (2019-07-01)
- "Use python3 explicitly" PR #26 by @jubalh merged
- inconsistency in the version numbering resolved

v1.1.1 (2019-06-30)
- launcher script for Void Linux added
- German description added to the .desktop file

v1.1.0 (2019-06-28)
- new: German translation by xsme
- new: sorting order -> settings, sorting order button & menu
- bug fix: column width 256 -> 280 to avoid the last column that does not fit in the screen
- bug fix: clear wallpaper selection on folder changed
- bug fix: do not activate the Trash button if image path not inside the home path

v1.0.0 (2019-06-25)
- missing check if button is not None added
- missing lang string added, lines sorted
- AUR - python-send2trash moved to optional dependencies

v0.1.3 (pre-release, 2019.06.24)
- PR #18 by Head-on-a-Stick merged - fix to issue #17 (Wallpaper unselectable w/o python3-send2trash (Debian))

v0.1.2 (pre-release, 2019.06.23)
- A try/except clause checks the python send2trash module availability.
  The python-send2trash dependency to be removed where unavailable.

v0.1.1 (pre-release, 2019.06.22)
- "azote/main.py: corrected feh options" PR #8 by Head-on-a-Stick merged
- some no longer needed print commands removed
- hardcoded strings replaced with the dictionary
- all modules removed from install_requires=[] in setup.py

v0.1.0 (pre-release, 2019.06.21)
- program icon slightly altered
- buttons and labels grouped in a single row
- new button: "Open selected with feh"
- new button: "Move selected to trash"
- new dependency: python-send2trash

v0.0.3 (pre-release, 2019.06.18):
- "azote/tools.py: death to grep!" PR by Head-on-a-Stick merged
- folder selection button - vertical size fixed
- deprecated GTK+ code brought up to date
- clearing background color selection on image split button pressed
- other minor fixes to UI

v0.0.2 (pre-release, 2019-06-17)
- shebangs changed to python3
- dependency on i3ipc-python dropped
- a bug manifesting on single-headed setup only (on changing the split_button sensitivity) fixed
- simple translations introduced

v0.0.1 (pre-release, 2019-06-15)
Initial release

azote/tools.py
#!/usr/bin/env python3
# _*_ coding: utf-8 _*_

"""
Wallpaper manager for Sway, i3 and some other WMs, as a frontend to swaybg and feh

Author: Piotr Miller
e-mail: nwg.piotr@gmail.com
Website: http://nwg.pl
Project: https://github.com/nwg-piotr/azote
License: GPL3
"""
import os
import glob
import hashlib
import logging
from PIL import Image, ImageOps
import common
import pickle
import subprocess
import locale
import pkg_resources
from shutil import copyfile

import json

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk


def log(message, level=None):
    if common.logging_enabled:
        if level == "critical":
            logging.critical(message)
        elif level == "error":
            logging.error(message)
        elif level == "warning":
            logging.warning(message)
        elif level == "info":
            logging.info(message)
        else:
            logging.debug(message)


def str_to_bool(s):
    if s.upper() == 'TRUE':
        return True
    elif s.upper() == 'FALSE':
        return False
    else:
        raise ValueError


def check_displays():
    # Sway or not Sway? If so, the `swaymsg -t get_seats` command should return exit code 0
    result = subprocess.run(['swaymsg', '-t', 'get_seats'], stdout=subprocess.DEVNULL)
    common.sway = result.returncode == 0
    # ask for the wm name (just for logging: maybe we should give up on it to drop the wmctrl dependency?)
    wm = subprocess.check_output("wmctrl -m | awk '/Name/{print $2}'", shell=True).decode("utf-8").strip()
    if common.sway:
        common.env['wm'] = 'sway'
    else:
        common.env['wm'] = wm

    fnull = open(os.devnull, 'w')
    common.env['xrandr'] = subprocess.call(["which", "xrandr"], stdout=fnull, stderr=subprocess.STDOUT) == 0

    if common.sway:
        # We need swaymsg to check outputs on Sway
        try:
            displays = []
            json_string = subprocess.check_output("swaymsg -t get_outputs", shell=True).decode("utf-8").strip()
            outputs = json.loads(json_string)
            for output in outputs:
                if output['active']:  # dunno WTF xroot-0 is: i3 returns such an output with "active":false
                    display = {'name': output['name'],
                               'x': output['rect']['x'],
                               'y': output['rect']['y'],
                               'width': output['rect']['width'],
                               'height': output['rect']['height']}
                    displays.append(display)
                    log("Output found: {}".format(display), common.INFO)

            # Dummy display for testing
            """display = {'name': 'HDMI-A-2',
                       'x': 1920,
                       'y': 0,
                       'width': 1920,
                       'height': 1080}
            displays.append(display)
            log("Output: {}".format(display), common.INFO)"""

            # sort displays list by x, y: from left to right, then from bottom to top
            displays = sorted(displays, key=lambda x: (x.get('x'), x.get('y')))

            return displays

        except Exception as e:
            log("Failed checking displays: {}".format(e))

    # On i3 we could use i3-msg here, but xrandr should also return what we need. If not on Sway - let's use xrandr
    elif common.env['xrandr']:
        names = subprocess.check_output("xrandr | awk '/ connected/{print $1}'", shell=True).decode(
            "utf-8").splitlines()
        res = subprocess.check_output("xrandr | awk '/*/{print $1}'", shell=True).decode("utf-8").splitlines()
        displays = []
        for i in range(len(names)):
            w_h = res[i].split('x')
            display = {'name': names[i],
                       'x': 0,
                       'y': 0,
                       'width': int(w_h[0]),
                       'height': int(w_h[1])}
            displays.append(display)
            log("Output found: {}".format(display), common.INFO)

        return displays

    else:
        log("Couldn't check displays", common.ERROR)
        exit(1)


def set_env(language=None):
    # application folder
    common.app_dir = os.path.join(os.getenv("HOME"), ".azote")
    if not os.path.isdir(common.app_dir):
        os.mkdir(common.app_dir)

    # logging
    common.log_file = os.path.join(common.app_dir, "log.txt")
    logging.basicConfig(filename=common.log_file, format='%(asctime)s %(levelname)s: %(message)s', filemode='w',
                        level=logging.INFO)

    try:
        version = pkg_resources.require(common.app_name)[0].version
    except Exception as e:
        version = ' unknown version: {}'.format(e)

    log('Azote v{}'.format(version), common.INFO)

    # We will preload the en_EN dictionary as default values
    common.lang = Language()

    if not language:
        # Lets check locale value
        # If running with LC_ALL=C, we'll get (None, None) here. Let's use en_EN in such case.
        lang = locale.getlocale()[0] if locale.getlocale()[0] is not None else 'en_EN'
    else:
        lang = language

    common.lang.load(lang)

    common.displays = check_displays()

    # thumbnails folder
    common.thumb_dir = os.path.join(common.app_dir, "thumbnails")
    if not os.path.isdir(common.thumb_dir):
        os.mkdir(common.thumb_dir)

    # command file
    common.cmd_file = os.path.join(os.getenv("HOME"), ".azotebg")

    # temporary folder
    common.tmp_dir = os.path.join(common.app_dir, "temp")
    if not os.path.isdir(common.tmp_dir):
        os.mkdir(common.tmp_dir)
    # remove all files inside on start
    for file in os.listdir(common.tmp_dir):
        path = os.path.join(common.tmp_dir, file)
        os.remove(path)  # clear on start
        log("Removed {}".format(path), common.INFO)

    # backgrounds folder
    name = 'backgrounds-sway' if common.sway else 'backgrounds-feh'
    common.bcg_dir = os.path.join(common.app_dir, name)
    if not os.path.isdir(common.bcg_dir):
        os.mkdir(common.bcg_dir)

    # Sample folder (will be set on 1st run only)
    common.sample_dir = os.path.join(common.app_dir, "sample")
    if not os.path.isdir(common.sample_dir):
        os.mkdir(common.sample_dir)
    copyfile('images/azote-wallpaper.jpg', os.path.join(common.sample_dir, 'azote-wallpaper.jpg'))
    copyfile('images/azote-wallpaper1.jpg', os.path.join(common.sample_dir, 'azote-wallpaper1.jpg'))
    copyfile('images/azote-wallpaper2.jpg', os.path.join(common.sample_dir, 'azote-wallpaper2.jpg'))

    # Sway comes with some sample wallpapers
    if common.sway and os.path.isdir('/usr/share/backgrounds/sway'):
        common.sample_dir = '/usr/share/backgrounds/sway'

    common.settings = Settings()

    log("Environment: {}".format(common.env), common.INFO)

    # check programs capable of opening files of allowed extensions
    if os.path.isfile('/usr/share/applications/mimeinfo.cache'):
        common.associations = {}  # Will stay None if the mimeinfo.cache file not found

        with open(os.path.join('/usr/share/applications/mimeinfo.cache')) as f:
            mimeinfo_cache = f.read().splitlines()

        for file_type in common.allowed_file_types:
            for line in mimeinfo_cache:
                if line.startswith('image/{}'.format(file_type)):
                    line = line.split('=')[1]  # cut out leading 'image/ext'
                    # Paths to .desktop files for opener names found
                    filenames = line[:-1].split(';')  # cut out trailing ';' to avoid empty last element after splitting
                    # prepend path
                    for i in range(len(filenames)):
                        filenames[i] = '/usr/share/applications/{}'.format(filenames[i])

                    data = []
                    for i in range(len(filenames)):
                        # Let's find the program Name= and Exec= in /usr/share/applications/shortcut_name.desktop
                        name, exe = '', ''

                        if os.path.isfile(filenames[i]):
                            with open(filenames[i]) as f:
                                rows = f.read().splitlines()
                                for row in rows:
                                    if row.startswith('Name='):
                                        name = row.split('=')[1]
                                    elif row.startswith('Name[{}]='.format(common.lang.lang[0:2])):
                                        name = row.split('=')[1]
                                    if row.startswith('Exec'):
                                        exe = row.split('=')[1].split()[0]
                                        continue
                            if name and exe:
                                data.append((name, exe))
                    common.associations[file_type] = data
                    """
                    Not necessarily all programs register jpg and jpeg extension (e.g. gimp registers jpeg only).
                    Let's create sets, join them and replace lists for both jpg and jpeg keys.
                    """
                    try:
                        jpg = set(common.associations['jpg'])
                        jpeg = set(common.associations['jpeg'])
                        together = jpg | jpeg
                        common.associations['jpg'] = together
                        common.associations['jpeg'] = together
                    except KeyError:
                        pass
        log("Image associations: {}".format(common.associations), common.INFO)
    else:
        print('Failed opening /usr/share/applications/mimeinfo.cache')
        log("Failed creating image associations: /usr/share/applications/mimeinfo.cache file not found."
            " Setting feh as the only viewer.", common.ERROR)


def copy_backgrounds():
    # Clear current folder content
    for file in os.listdir(common.bcg_dir):
        os.remove(os.path.join(common.bcg_dir, file))
    # Copy manipulated (flip, split) files from the temporary folder
    for file in os.listdir(common.tmp_dir):
        copyfile(os.path.join(common.tmp_dir, file), os.path.join(common.bcg_dir, file))


def hash_name(full_path):
    """
    Thumbnail path -> name (w/o extension)
    :param full_path: original file path
    :return: MD5-hashed path
    """
    return hashlib.md5(full_path.encode()).hexdigest()


def create_thumbnails(scr_path):
    # Let's count allowed files first
    common.progress_bar.hide()
    counter = 0
    for extension in common.allowed_file_types:
        for in_path in glob.glob(os.path.join(scr_path, "*.{}".format(extension))):
            if file_allowed(in_path):
                counter += 1
    if counter > 0:
        # get all the files of allowed types from current path
        common.progress_bar.show()
        common.progress_bar.set_fraction(0.0)
        processed = 0
        for extension in common.allowed_file_types:
            for in_path in glob.glob(os.path.join(scr_path, "*.{}".format(extension))):
                if file_allowed(in_path):
                    thumb_name = "{}.png".format(hash_name(in_path))
                    dest_path = os.path.join(common.thumb_dir, thumb_name)
                    if not os.path.isfile(dest_path):
                        create_thumbnail(in_path, dest_path, thumb_name)
                    elif is_newer(in_path, dest_path):
                        create_thumbnail(in_path, dest_path, thumb_name, True)
                    processed += 1
                    common.progress_bar.set_fraction(processed / counter)
                    common.progress_bar.set_text(str(processed))
                    while Gtk.events_pending():
                        Gtk.main_iteration()
    common.progress_bar.hide()


def create_thumbnail(in_path, dest_path, thumb_name, refresh=False):
    action = 'New thumb' if not refresh else 'Refresh'
    try:
        img = Image.open(in_path)
        # convert to thumbnail image
        img.thumbnail((240, 135), Image.ANTIALIAS)

        img = expand_img(img)

        img.save(dest_path, "PNG")
        log('{}: {} -> {}'.format(action, in_path, thumb_name), common.INFO)
    except Exception as e:
        log('{} - {}'.format(action, e), common.ERROR)


def flip_selected_wallpaper():
    """
    This creates vertically flipped image and its thumbnail and saves to ~/.azote/backgrounds
    :return: thumbnail path, flipped image path
    """
    if common.selected_wallpaper:
        try:
            img = Image.open(common.selected_wallpaper.source_path)
            flipped = img.transpose(Image.FLIP_LEFT_RIGHT)
            img_path = os.path.join(common.bcg_dir, "flipped-{}".format(common.selected_wallpaper.filename))
            flipped.save(os.path.join(common.tmp_dir, "flipped-{}".format(common.selected_wallpaper.filename)), "PNG")

            flipped.thumbnail((240, 135), Image.ANTIALIAS)
            flipped = expand_img(flipped)

            thumb_path = os.path.join(common.tmp_dir, "thumbnail-{}".format(common.selected_wallpaper.filename))
            flipped.save(thumb_path, "PNG")
            return thumb_path, img_path

        except Exception as e:
            log('Failed flipping {} - {}'.format(common.selected_wallpaper.source_path, e), common.ERROR)


def split_selected_wallpaper(num_parts):
    try:
        img = Image.open(common.selected_wallpaper.source_path)
        width, height = img.size
        part_width = width // num_parts
        paths_list = []
        for i in range(num_parts):
            box = (i * part_width, 0, i * part_width + part_width, height)
            part = img.crop(box)
            img_path = os.path.join(common.bcg_dir, "part{}-{}".format(i, common.selected_wallpaper.filename))
            part.save(os.path.join(common.tmp_dir, "part{}-{}".format(i, common.selected_wallpaper.filename)), "PNG")

            part.thumbnail((240, 135), Image.ANTIALIAS)

            thumb_path = os.path.join(common.tmp_dir, "thumb-part{}-{}".format(i, common.selected_wallpaper.filename))

            part = expand_img(part)

            part.save(thumb_path, "PNG")
            paths = (img_path, thumb_path)
            paths_list.append(paths)
        return paths_list

    except Exception as e:
        log('Failed splitting {} - {}'.format(common.selected_wallpaper.source_path, e), common.ERROR)


def expand_img(image):
    # We want the thumbnail to be always 240 x 135. Let's expand if necessary.
    width, height = image.size
    border_h = (240 - width) // 2
    border_v = (135 - height) // 2
    if border_v > 0 or border_h > 0:
        border = (border_h, border_v, border_h, border_v)
        return ImageOps.expand(image, border=border)
    else:
        return image
    
    
def scale_and_crop(item, image_path, display):
    img = Image.open(image_path)
    
    # We can either scale vertically & crop horizontally or scale horizontally and crop vertically
    new_height = int(img.size[0] * display['height'] / display['width'])

    if new_height < img.size[1]:  # we need to scale to display width and crop vertical margins
        new_height = int(display['width'] * img.size[1] / img.size[0])
        # Choose the filter depending on if we're scaling down or up
        if new_height >= display['height']:
            img = img.resize((display['width'], new_height), Image.ANTIALIAS)
        else:
            img = img.resize((display['width'], new_height), Image.BILINEAR)

        margin = (img.size[1] - display['height']) // 2
        img = img.crop((0, margin, display['width'], display['height'] + margin))

    elif new_height > img.size[1]:  # we need to scale to display height and crop horizontal margins
        new_width = int(img.size[0] * display['height'] / img.size[1])
        if new_width >= display['width']:
            img = img.resize((new_width, display['height']), Image.ANTIALIAS)
        else:
            img = img.resize((new_width, display['height']), Image.BILINEAR)

        margin = (img.size[0] - display['width']) // 2
        img = img.crop((margin, 0, display['width'] + margin, display['height']))

    else:
        img = img.resize((display['width'], display['height']), Image.ANTIALIAS)

    img.save('{}-{}x{}{}'.format(os.path.splitext(image_path)[0], display['width'], display['height'], os.path.splitext(image_path)[1]))
    common.preview.refresh()


def is_newer(in_path, dest_path):
    return os.path.getmtime(in_path) > os.path.getmtime(dest_path)


def file_allowed(path):
    ext = path.split('.')[-1].lower()
    return ext in common.allowed_file_types


def update_status_bar():
    num_files = 0
    total_size = 0
    if os.path.isdir(common.thumb_dir):
        for file in os.listdir(common.thumb_dir):
            num_files += 1
            file_info = os.stat(os.path.join(common.thumb_dir, file))
            total_size += file_info.st_size
    common.status_bar.push(0, common.lang['thumbnails_in_cache'].format(num_files, convert_bytes(total_size)))


def convert_bytes(num):
    """
    https://stackoverflow.com/a/39988702/4040598
    """
    for x in ['bytes', 'KB', 'MB', 'GB', 'TB']:
        if num < 1024.0:
            return "%3.1f %s" % (num, x)
        num /= 1024.0


def rgba_to_hex(color):
    """
    Return hexadecimal string for :class:`Gdk.RGBA` `color`
    http://wrhansen.blogspot.com/2012/09/how-to-convert-gdkrgba-to-hex-string-in.html
    """
    return "#{0:02x}{1:02x}{2:02x}".format(int(color.red * 255),
                                           int(color.green * 255),
                                           int(color.blue * 255))


class Settings(object):
    def __init__(self):
        self.file = os.path.join(common.app_dir, "settings.pkl")

        self.src_path = common.sample_dir
        self.sorting = 'new'
        # Gtk.Menu() on sway is unreliable, especially called with right click
        self.show_open_button = common.sway         # shown by default on Sway
        self.show_context_menu = not common.sway    # turned off by default on Sway

        self.load()

    def load(self):
        save_needed = False
        if not os.path.isfile(self.file):
            log('Creating initial settings', common.INFO)
            self.save()

        with open(self.file, 'rb') as input_data:
            settings = pickle.load(input_data)

        self.src_path = settings.src_path  # holds selected path to source pictures
        try:
            self.sorting = settings.sorting  # 'new' 'old' 'az' 'za'
            log('Image sorting: {}'.format(self.sorting), common.INFO)
        except AttributeError:
            save_needed = True
            
        try:
            self.show_open_button = settings.show_open_button
            log('Picture menu button: {}'.format(self.show_open_button), common.INFO)
        except AttributeError:
            save_needed = True

        try:
            self.show_context_menu = settings.show_context_menu
            log('Thumbnail context menu: {}'.format(self.show_context_menu), common.INFO)
        except AttributeError:
            save_needed = True
            
        if save_needed:
            self.save()

    def save(self):
        with open(self.file, 'wb') as output:
            pickle.dump(self, output, pickle.HIGHEST_PROTOCOL)


class Language(dict):
    def __init__(self):
        super().__init__()
        self.lang = 'en_EN'
        # We'll initialize with values from en_EN
        with open(os.path.join('languages', 'en_EN')) as f:
            lines = f.read().splitlines()
            for line in lines:
                if line and not line.startswith('#'):
                    pair = line.split('=')
                    key, value = pair[0].strip(), pair[1].strip()
                    self[key] = value

    def load(self, lang):
        try:
            # Overwrite initial values if translation found
            with open(os.path.join('languages', lang)) as f:
                lines = f.read().splitlines()
                for line in lines:
                    if line and not line.startswith('#'):
                        pair = line.split('=')
                        key, value = pair[0].strip(), pair[1].strip()
                        self[key] = value
            self.lang = lang
            log("Loaded lang: {}".format(lang), common.INFO)

        except FileNotFoundError:
            log("Couldn't load lang: {}".format(lang), common.WARNING)
